### ISSUE TITLE
Performance issue (spikes in utility processors)

### ISSUE THREAD

**Issue by bandokihiro** [StanfordLegion/legion#1203]
I am encountering performance issues that create gaps in my GPU timeline and prevent my application from strong scaling. After eliminating side issues that could affect performance, I think I narrowed it down to spikes in the utility processor. They are regularly spaced and, when they occur, some rank has an unusually long task box. Such gap then propagates to the rest of the system. I have seen `Post-Task Execution` taking O(10) [ms]. I have also seen light green boxes in the utility processor that don't display anything when I hover over them.

The following [profile](http://sapling.stanford.edu/~bandokihiro/6_02/legion_prof) is a 2-node 6-GPU-per-node run on Summit. The command line was
```
jsrun -n 12 -r 6 -a 1 -c 7 -g 1 -b packed:7 -l gpu-cpu /gpfs/alpine/scratch/bandok/csc335/Softwares/DG-Legion/build_1/exec/solver -logfile logs/log_%.log -hdf5:forcerw -dm:memoize -ll:cpu 0 -ll:util 2 -ll:bgwork 3 -ll:bgworkpin 1 -ll:csize 50000 -ll:rsize 0 -ll:gpu 1 -ll:fsize 14000 -ll:onuma 0 -ll:ocpu 1 -ll:othr 3 -lg:prof 12 -lg:prof_logfile prof_%.gz -ll:ib_rsize 512m -ll:ib_zsize 0 -ll:ext_sysmem 0
```
There are 3 pinned bg workers and 3 pinned omp threads which leaves 1 core for the utility. To easily visualize the problem, please follow these steps:
1. select only GPU in the profiling options
2. visualize all GPU timelines (12 ranks)
3. each timeline starts with a long light green box; zoom from there to the end of the timeline
 
At this point, the gaps should be visible. If you deploy the utility timelines, you can see that, at every gap, all utility procs' utilization spikes. For every gap, there is one rank with an unusually long box in the GPU timeline. I usually look at what's happening in the utiliy for the same rank.

Any help would be greatly appreciated.

**Comment by bandokihiro** [comment]
One thing that is probably relevant to mention is that there is a task that is launched over a sparse 2D domain. Each rank i launches a point task "for all neighboring rank j" to pack local data that need to be sent to them @streichler mentioned when I started that he wasn't sure what the overhead of such launch would be.

**Comment by bandokihiro** [comment]
Here is the new [profile](http://sapling.stanford.edu/~bandokihiro/6_02_2) obtained with the most recent CR. The light green boxes in the utility processors that had missing metadata are "Replay Phycal Trace" tasks. Gaps tend to happen near a spike in the utility and these replay tasks can take several ms. I ran with `-lg:warn` and no warning was generated. Since I regularly check this, I silence warnings at most of the future `get_result` calls but I am relatively confident that this run (in which I commented a lot of stuff related to futures) doesn't have any wait on futures.

**Comment by lightsighter** [comment]
Now that we can see the boxes, we can see that the runtime trace replay is way ahead of the execution as we would expect, so the problem is not waits.

The problem here is almost certainly that there is interference between the utility processor threads and the background worker threads the service the GPUs. 

> There are 3 pinned bg workers and 3 pinned omp threads which leaves 1 core for the utility.

If that's the case why run with two utility processors? Why not just one?

What does the core mapping look like in each process? If you running with multiple processes/node, how are you making sure that those processes are mapping onto different cores?

**Comment by bandokihiro** [comment]
> What does the core mapping look like in each process? If you running with multiple processes/node, how are you making sure that those processes are mapping onto different cores?

The mapping is made by the `jsrun` command. [Here](https://jobstepviewer.olcf.ornl.gov/summit/1877226-1/h17n08) is a visualization (with `smt1`). Each GPU/rank is mapped to 7 different cores.

> If that's the case why run with two utility processors? Why not just one?

When I started, I stumbled upon a situation where 1 utility processor was not enough. Since then, using 2 utility processsors is kind of my default. In my main branch, I don't specify anything about background workers. When I did the implicit ghost portable version, copies were a bottleneck and, with @streichler, we saw improvements with 3 bg workers per rank that are pinned. 

I have done 3 additoinal runs:

- with `-ll:util 1 -ll:bgwork 2 -ll:bgworkpin 1` ([profile](http://sapling.stanford.edu/~bandokihiro/6_02_3))
- with `-ll:util 2 -ll:bgwork 2 -ll:bgworkpin 1` ([profile](http://sapling.stanford.edu/~bandokihiro/6_02_4))
- with `-ll:util 2 -ll:bgwork 2 -ll:bgworkipin 1` and `smt4` ([profile](http://sapling.stanford.edu/~bandokihiro/6_02_5), [core map](https://jobstepviewer.olcf.ornl.gov/summit/1878758-1/g34n18))

They all more or less show the same symptoms. Gaps are not necessarilly located at the spikes in the utility, but there is a qualitative correlation between the two and, often time, there is a rank that has a gap at the spike.

As to the mapping with `smt4`, `-ll:show_rsrv` gave this
```
     dedicated worker (generic) #1: allocated <0>
     dedicated worker (generic) #2: allocated <4>
     dedicated worker (generic) #3: allocated <8>
     utility proc 1d00000000000000: allocated <24,25,26,27>
     utility proc 1d00000000000001: allocated <24,25,26,27>
     OMP-1 proc 1d00000000000002 (master): allocated <12>
     OMP-1 proc 1d00000000000002 (worker 1): allocated <16>
     OMP-1 proc 1d00000000000002 (worker 2): allocated <20>
     GPU ctxsync 0x2b1c3e30: allocated <24,25,26,27>
     GPU proc 1d00000000000003: allocated <24,25,26,27>
     GPU ctxsync 0x3d0c0050: allocated <52,53,54,55>
     GPU proc 1d00010000000003: allocated <52,53,54,55>
```
There is 21 cores availble on each socket. If I assign 7 cores to a rank, realm sees cores 0-27. It seems that the bg workers and the omp threads that are pinned occupy an entire core. All 4 threads of the 7th core appear individually for the other threads, including the utility.

**Comment by bandokihiro** [comment]
This [profile](http://sapling.stanford.edu/~bandokihiro/6_04_2/legion_prof) obtained for a 4-node run, `smt4`, and `-ll:util 2 -ll:bgwork 2 -ll:bgworkpin 1` shows a completely different utilty behavior. The stalls in the GPU timelines tend to happen on all GPUs of a node at the same time. This profile, which we have seen before, shows this behavior very clearly.

**Comment by bandokihiro** [comment]
In an effort to simplify the case, this [profile](http://sapling.stanford.edu/~bandokihiro/2_01_01) is a 1-node 2-rank run, with `smt4`, and 21 cores per rank, i.e. a ton of resources for the threads created by Legion/Realm. Here are my obervations:

- there are still unusually long tasks that don't necessarilly happen at the spikes in the utility
- when there is a spike in the utility, there is also something happening on the OMP proc which is used as the main application CPU, is that expected?

Below is the variation of the iiteration wall-time. In order to minimize any potential overhead, I am writing these to a global array which I dump only at the end.
![Figure 11](https://user-images.githubusercontent.com/39278362/156264595-3a2a1dad-dc10-4068-8f20-e7b6a9f81acc.png)

**Comment by bandokihiro** [comment]
I think I found a solution. Currently working on it.

**Comment by bandokihiro** [comment]
You are correct, and the spikes in the utility were a misleading observation, as were a lot of other things when the time scale of an iteration is O(10) ms. A lot of things can induce small perturbations and it took me some time to narrow the real issue down. Long story short, I was able to get rid of these anomalies by:

- using `smt4` instead of `smt1`. Indeed, 7 cores are just not enough for the 14 threads that are actually created by each process. In my experience, classical hyper-threading with intel processors did not improve the performance of the application threads but I am not very familiar with the architecture on Summit. Moreover, maybe hyper-threading (or its version on Summit) is beneficial for the runtime threads
- using `-cuda:legacysync 1` (introduced in 90fba2ec). My reasoning for this is that the main difference between MPI+Kokkos and Legion+Kokkos is that in the former case, if not doing anything fancy with CUDA streams, only the default one is used. With Legion+Kokkos, Realm submits for instance d2d copies on a different stream and it seems that it interfers with the application kernels.

Using only one of the above is not sufficient. With both, I get a strong scaling up to 8-node that is close to optimal from just measuring by eye what an ideal iteration time would be if discarding the gaps.

To scale further (weak or strong), I bump into #1193. This is reproducible and fails every time.

This seems more than a hack than anything else to me. But it works. I'll let @streichler comment on the interference when using multiple streams.

**Comment by magnatelee** [comment]
> using smt4 instead of smt1.

FWIW, smt4 was the default mode that I ran my jobs in on Power-based systems. My theory was that the threads we get with smt1 have diminished hardware resources, which make them not capable of running the runtime code fast enough. (This also seems to suggest that those threads are more cleanly isolated at the hardware level than Intel's hyperthreads, which effectively share most of the resource on each core.)

**Comment by lightsighter** [comment]
Not having used PowerPC chips much, I'll defer to Wonchan's experience here. SMT is a fickle thing and means different things to different hardware vendors. They all have different ways of doing it and with different constraints. I wouldn't necessarily consider it a hack to change the SMT setting. IBM put it there for a reason, because clearly they knew it was something that different users would want control over. Not all threads are created equal and some programs will push on the hardware harder than others.

**Comment by bandokihiro** [comment]
Thank you all for your comments. I met with @streichler today and he said that it is not a hack to have to use that flag. Another thing that was mentioned is that a better solution would be to tell Kokkos not to use the default stream but the one "used for the task" (I am not sure that statement is correct). In the next days, I am gonna run my application on PizDaint to see if this is specific to Power-based systems. Once I have the results for this, I am gonna close the issue.
